### PR TITLE
superuser support and domain based org autoinvite

### PR DESCRIPTION
**Pull request by pjain1**
Resolves https://github.com/rilldata/rill-developer/issues/2061

### CODE REVIEW

File: cli/cmd/root.go [hunk 1]
```diff
@@ -17,6 +17,7 @@ import (
 	"github.com/rilldata/rill/cli/cmd/runtime"
 	"github.com/rilldata/rill/cli/cmd/source"
 	"github.com/rilldata/rill/cli/cmd/start"
+	"github.com/rilldata/rill/cli/cmd/superuser/autoinvite"
```
Comment (begelundmuller) [inline cli/cmd/root.go:20]: The idea of a superuser namespace is nice, let's make it more explicit. Suggested refactors:

1. Rename `.../cmd/superuser` to `.../cmd/sudo`
2. Mount a `SudoCmd` on `rill sudo`, and make it `Hidden: true`
3. Then put the auto invite commands under that, like `rill sudo autoinvite add`

File: cli/cmd/superuser/autoinvite/create.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+	"github.com/rilldata/rill/cli/pkg/cmdutil"
+	"github.com/rilldata/rill/cli/pkg/config"
+	adminv1 "github.com/rilldata/rill/proto/gen/rill/admin/v1"
+	"github.com/spf13/cobra"
+)
+
+func CreateCmd(cfg *config.Config) *cobra.Command {
+	createCmd := &cobra.Command{
+		Use:    "create <org> <domain> <role>",
```
Comment (begelundmuller) [inline cli/cmd/superuser/autoinvite/create.go:14]: nit: I think in other commands we pair `create/delete` and `add/remove` – so either change this to `add` or change the remove command to `delete`

File: cli/cmd/superuser/autoinvite/create.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+		Args:   cobra.ExactArgs(3),
+		Short:  "Autoinvite users from a domain",
+		Hidden: true,
```
Comment (begelundmuller) [inline cli/cmd/superuser/autoinvite/create.go:17]: Remove `Hidden` here and put it on the `sudo` command instead (then doing `rill sudo autoinvite -h` will show the available commands)

File: cli/cmd/superuser/autoinvite/autoinvite.go [hunk 4]
```diff
@@ -0,0 +1,17 @@
+package autoinvite
+
+import (
+	"github.com/rilldata/rill/cli/pkg/config"
+	"github.com/spf13/cobra"
+)
+
+func AutoinviteCmd(cfg *config.Config) *cobra.Command {
+	autoinviteCmd := &cobra.Command{
+		Use:   "auto-invite",
```
Comment (begelundmuller) [inline cli/cmd/superuser/autoinvite/autoinvite.go:10]: Make it `autoinvite` (no dash)

**Review by begelundmuller — CHANGES_REQUESTED** [review]

File: admin/database/validate.go [hunk 5]
```diff
@@ -17,6 +17,9 @@ var validate *validator.Validate
 // slugRegexp is used to validate identifying names (e.g. "rill-data", not "Rill Data").
 var slugRegexp = regexp.MustCompile("^[_a-zA-Z0-9][-_a-zA-Z0-9]{2,39}$")
 
+// Regular expression to match a domain name
+var domainRegex = regexp.MustCompile(`^[a-zA-Z0-9]+([-.][a-zA-Z0-9]+)*\.[a-zA-Z]{2,}$`)
```
[the remark this answers is not in the document]
Comment (pjain1) [reply admin/database/validate.go:21]: both `.` and `\.` will match character `.` as they are part of the list in square brackets, whats the difference ?

File: admin/server/auth/claims.go [hunk 6]
```diff
@@ -76,6 +81,7 @@ type authTokenClaims struct {
 	admin                   *admin.Service
 	orgPermissionsCache     map[string]*adminv1.OrganizationPermissions
 	projectPermissionsCache map[string]*adminv1.ProjectPermissions
+	superuser               TriBool
```
[the remark this answers is not in the document]
Comment (pjain1) [reply admin/server/auth/claims.go:84]: yeah was wondering about it, thanks

File: admin/user.go [hunk 7]
```diff
@@ -77,6 +78,29 @@ func (s *Service) CreateOrUpdateUser(ctx context.Context, email, name, photoURL
 		}
 	}
 
+	// check if users email domain is in autoinvite list
+	if len(orgInvites) == 0 {
```
[the remark this answers is not in the document]
Comment (pjain1) [reply admin/user.go:82]: I think we can just process `orgInvites` first and for autoinvites if user is already member of the org skip it.

File: admin/user.go [hunk 8]
```diff
@@ -77,6 +78,29 @@ func (s *Service) CreateOrUpdateUser(ctx context.Context, email, name, photoURL
...
+		domain := email[strings.Index(email, "@")+1:]
```
[the remark this answers is not in the document]
Comment (pjain1) [reply admin/user.go:83]: so many rules

File: admin/database/validate.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (begelundmuller) [reply admin/database/validate.go:21]: My mistake, didn't realize `.` inside square brackets is not interpreted as any character

File: admin/user.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (begelundmuller) [reply admin/user.go]: Haha

**Pull request merged**